digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/process/SpeedLimitValidator.scala

```diff
@@ -0,0 +1,114 @@
+    val speedLimitGeometry = GeometryUtils.truncateGeometry2D(roadLink.geometry, speedLimit.startMeasure, speedLimit.endMeasure)
+    val (first, last) = GeometryUtils.geometryEndpoints(speedLimitGeometry)
+
+    val trafficSingsByRadius =
+          (speedLimit.sideCode match {
+            case SideCode.TowardsDigitizing =>
+              trafficSignService.getTrafficSignByRadius(first, 50, Some(SpeedLimits)).filter (_.validityDirection == speedLimit.sideCode.value)
+
+            case SideCode.AgainstDigitizing =>
+              trafficSignService.getTrafficSignByRadius(last, 50, Some(SpeedLimits)).filter (_.validityDirection == speedLimit.sideCode.value)
+
+            case _ =>
+              roadLink.trafficDirection match {
+                case TrafficDirection.TowardsDigitizing =>
+                  trafficSignService.getTrafficSignByRadius(first, 50, Some(SpeedLimits))
+                    .filter (trafficSign => SideCode.apply(trafficSign.validityDirection) == TrafficDirection.toSideCode(TrafficDirection.TowardsDigitizing))
+
+                case TrafficDirection.AgainstDigitizing =>
+                  trafficSignService.getTrafficSignByRadius(last, 50, Some(SpeedLimits))
+                    .filter (trafficSign => SideCode.apply(trafficSign.validityDirection) == TrafficDirection.toSideCode(TrafficDirection.AgainstDigitizing))
+
+                case _ =>
+              trafficSignService.getTrafficSignByRadius(first, 50, Some(SpeedLimits)).filter(_.validityDirection == SideCode.TowardsDigitizing.value) ++
+                trafficSignService.getTrafficSignByRadius(last, 50, Some(SpeedLimits)).filter(_.validityDirection == SideCode.AgainstDigitizing.value)
+
+          }}).filter(_.linkId == speedLimit.linkId)
+
+    if (trafficSingsByRadius.nonEmpty) {
+      (trafficSingsByRadius.groupBy(minDistance(_, first)) ++ trafficSingsByRadius.groupBy(minDistance(_, last))).reduceLeft(min)._2
+    }else
+      Seq()
+  }
+
+  private def speedLimitValueValidator(speedLimit: SpeedLimit, trafficSign: PersistedTrafficSign): Option[SpeedLimit] = {
+    val startUrbanAreaSpeedLimit: Int = 50
+    val endUrbanAreaSpeedLimit: Int = 80
+
+    speedLimit.value match {
+      case Some(NumericValue(speedLimitValue)) =>
+        TrafficSignType.apply(trafficSign.propertyData.find(p => p.publicId == "trafficSigns_type").get.values.head.propertyValue.toInt) match {
+          case TrafficSignType.SpeedLimit | TrafficSignType.SpeedLimitZone =>
+            trafficSign.propertyData.find(p => p.publicId == "trafficSigns_value").get.values.headOption match {
+              case Some(trafficSignValue) if trafficSignValue.propertyValue != speedLimitValue.toString =>
+                Some(speedLimit)
+              case _ => None
+            }
+
+          case TrafficSignType.EndSpeedLimitZone | TrafficSignType.EndSpeedLimit =>
+            trafficSign.propertyData.find(p => p.publicId == "trafficSigns_value").get.values.headOption match {
```

Comment:
And if we used the modifications above, in this line we only need to call getTrafficSignType function again to return the right TrafficSignType.